src/cache.rs

```diff
@@ -5,130 +5,111 @@
-    ///
-    /// # Panics
-    /// If called more than once or if unable to remove old cache file.
-    pub(crate) fn set_data_path(data_dir: PathBuf) -> Result<(), Error> {
-        // remove old cache if it's still there
-        let db_path = data_dir.join("cache.db");
-        if db_path.exists()
-            && !db_path.is_dir()
-            && fs::remove_file(&db_path).is_err()
-        {
-            panic!("Failed to remove old cache files. Please remove {} manually and try again.", db_path.display())
+    /// Returns a new cache instance.
+    pub(crate) fn new<T: AsRef<Path>>(path: T) -> Result<Self, Error> {
+        let db: DB;
+        let mut opts = Options::default();
+        opts.create_if_missing(true);
+        opts.create_missing_column_families(true);
+        // After 10 million bytes, sort the cache file and create new one
+        opts.set_write_buffer_size(10_000_000);
+
+        let list = DB::list_cf(&Options::default(), &path);
+
+        if list.is_err() {
+            db = DB::open(&opts, path)?;
+        } else {
+            db = DB::open_cf(&opts, path, list?)?;
         }
 
-        CACHE_DATA_PATH
-            .set(data_dir)
-            .expect("cache path can only be set once");
-
-        Persistence::with_backend(&backend(), |_| Ok(()))?;
-
-        Ok(())
-    }
-
-    /// Persist the cache to the internal backend.
-    pub(crate) fn persist(&self) -> Result<(), Error> {
-        let data_dir = CACHE_DATA_PATH.get().expect("cache path must be set");
-        let id_path = data_dir.join("cache.id");
-
-        let persistence_id = Persistence::persist(&backend(), &self.data)?;
-
-        let mut bytes = [0u8; 36];
-        let mut sink = Sink::new(&mut bytes);
-
-        persistence_id.encode(&mut sink);
+        let write_batch = WriteBatch::default();
 
-        fs::write(id_path, bytes)?;
-
-        Ok(())
+        Ok(Self { db, write_batch })
     }
 
-    /// Insert a note into the cache at the given block height. If note is
-    /// `None` only the block height gets updated.
+    // We store a column family named by hex representation of the psk.
+    // We store key as (psk, note) and value as (KeyData)
+    // The Canon of the tuple, we use it so we can have unique keys per psk and
+    // note
     pub(crate) fn insert(
         &mut self,
         psk: PublicSpendKey,
         height: u64,
         note: Option<Note>,
     ) -> Result<(), Error> {
-        if self.data.get(&psk)?.is_none() {
-            self.data.insert(
-                psk,
-                KeyData {
-                    last_height: 0,
-                    notes: BTreeSet::new(),
-                },
-            )?;
-        }
+        let cf: &ColumnFamily;
+        let cf_name = format!("{:?}", psk);
 
-        let mut key_data = self.data.get_mut(&psk)?.expect("psk");
+        let mut data = KeyData {
+            last_height: height,
+            notes: BTreeSet::new(),
+        };
```

Comment:
Since you moved last_height outside of the data (and I like it), I'm for removing the `KeyData` at all, and just store the `BtreeSet`

Reply:
Storing `(NoteHash, NoteData)` now